wireapp/wire-ios-mono
# fix: Ensure cookie is decryptable when checking if user logged in - WPB-10462

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10462" title="WPB-10462" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-10462</a>  [iOS] Issues around login & logout behavior
  </td></table>
  <br />
 







































### Issue

There are issues when logging out and logging in. From the ticket:

> It Iooks like data wasn't wiped and the app behaves erratically. Sometimes getting session expired, sometimes getting a prompt to enter password again which disappears

This behavior is similar to that of when a cookie expires - which results in erratic behavior. (There is a separate ticket for this).

I have not been able to reproduce the issue. However, we have a lot of code that checks whether a user is logged in by seeing if there is code in the keychain for this user. This is bad because:

1. Data from keychain isn't deleted when the app is deleted. Instead, for security reasons, it is encrypted with a key stored in user defaults. As user defaults is deleted when the app is deleted, while cookie data remains between and app uninstall and re-install cycle, that data is no longer readable so we should no longer use it to determine if a user is logged in.
2. Even if the data is readable it might be expired and so also shouldn't be used to determine if a user is logged in.

I originally addressed both of the above points, so that any time we check whether a user is logged in which check that the cookie exists, can be decrypted and hasn't expired. However I think that was too aggressive a change - perhaps some things only need to know that the cookie exists not that it is still valid.

Therefore this PR addresses point `1` because a cookie that can't be decrypted implies that it was added in a previous install and the app has been uninstalled and reinstalled.

Specifically this PR:

- Creates a single implementation to check whether a user has a cookie - `hasAuthenticationCookie` which checks that the cookie data exists and can be decrypted. Previously we had many implementations which all just checked whether the cookie data exists.
- Makes many changes to tests so that we pass valid cookie data instead of fake data.
- Introduces a new framework, `SwiftTransportSupport`, with some shared code for creating cookie data for tests.

### Testing

#### Test 1

1. Log in
2. Delete the app
3. Reinstall
4. You should be logged out
5. Try logging in and out repeatably
